src/ranch.erl

```diff
@@ -197,16 +200,42 @@ child_spec(Ref, Transport, TransOpts0, Protocol, ProtoOpts) ->
 		Ref, Transport, TransOpts, Protocol, ProtoOpts
 	]}, type => supervisor}.
 
--spec handshake(ref()) -> {ok, ranch_transport:socket()}.
+-spec handshake(ref()) -> {ok, ranch_transport:socket()} | {continue, any()}.
 handshake(Ref) ->
-	handshake(Ref, []).
+	handshake1(handshake, handshake, Ref, undefined).
 
--spec handshake(ref(), any()) -> {ok, ranch_transport:socket()}.
+-spec handshake(ref(), any()) -> {ok, ranch_transport:socket()} | {continue, any()}.
 handshake(Ref, Opts) ->
-	receive {handshake, Ref, Transport, CSocket, HandshakeTimeout} ->
-		case Transport:handshake(CSocket, Opts, HandshakeTimeout) of
+	handshake1(handshake, handshake, Ref, {opts, Opts}).
+
+-spec handshake_continue(ref()) -> {ok, ranch_transport:socket()} | {continue, any()}.
+handshake_continue(Ref) ->
+	handshake1(handshake_continue, handshake_continue, Ref, undefined).
+
+-spec handshake_continue(ref(), any()) -> {ok, ranch_transport:socket()}.
+handshake_continue(Ref, Opts) ->
+	handshake1(handshake_continue, handshake_continue, Ref, {opts, Opts}).
+
+-spec handshake_cancel(ref()) -> any().
+handshake_cancel(Ref) ->
+	receive {handshake_continue, Ref, Transport, CSocket, _} ->
+		Transport:handshake_cancel(CSocket)
+	end.
+
+handshake1(ExpectedMsg, HandshakeFun, Ref, Opts) ->
+	receive {ExpectedMsg, Ref, Transport, CSocket, HandshakeTimeout} ->
+		HandshakeResult = case Opts of
+			undefined ->
+				Transport:HandshakeFun(CSocket, HandshakeTimeout);
+			{opts, Opts1} ->
+				Transport:HandshakeFun(CSocket, Opts1, HandshakeTimeout)
+		end,
+		case HandshakeResult of
```

Comment:
handshake1 should just be this case clause.

Reply:
Ok, can do, but then I'll have to repeat all the other stuff around it in 4 functions, `handshake/1,2` and `handshake_continue/1,2`. If that's fine with you, then sure, fine with me :)

Reply:
I mean you can probably have `handshake/1` and `handshake_continue/1` share a function, the other two another, and then one function to check the result.